ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/LogAppendingCommitStrategy.scala

```diff
@@ -9,48 +9,46 @@ import com.daml.ledger.participant.state.kvutils.DamlKvutils.{
   DamlStateKey,
   DamlStateValue
 }
+import com.daml.ledger.participant.state.kvutils.Envelope
 import com.daml.ledger.participant.state.kvutils.export.SubmissionAggregator
-import com.daml.ledger.participant.state.kvutils.{Envelope, `Bytes Ordering`}
 import com.daml.ledger.participant.state.v1.ParticipantId
-import com.daml.ledger.validator.LedgerStateOperations.{Key, Value}
 
-import scala.collection.{SortedMap, breakOut}
 import scala.concurrent.{ExecutionContext, Future}
 
 class LogAppendingCommitStrategy[Index](
     ledgerStateOperations: LedgerStateOperations[Index],
     keySerializationStrategy: StateKeySerializationStrategy,
 )(implicit executionContext: ExecutionContext)
     extends CommitStrategy[Index] {
+  private val stateSerializationStrategy = new StateSerializationStrategy(keySerializationStrategy)
+
   override def commit(
       participantId: ParticipantId,
       correlationId: String,
       entryId: DamlLogEntryId,
       entry: DamlLogEntry,
       inputState: Map[DamlStateKey, Option[DamlStateValue]],
       outputState: Map[DamlStateKey, DamlStateValue],
-      exporterWriteSet: Option[SubmissionAggregator.WriteSetBuilder] = None,
-  ): Future[Index] = {
-    val serializedKeyValuePairs: SortedMap[Key, Value] =
-      outputState
-        .map {
-          case (key, value) =>
-            (keySerializationStrategy.serializeStateKey(key), Envelope.enclose(value))
-        }(breakOut)
-    exporterWriteSet.foreach {
-      _ ++= serializedKeyValuePairs
-    }
+      writeSetBuilder: Option[SubmissionAggregator.WriteSetBuilder] = None,
+  ): Future[Index] =
     for {
-      _ <- if (serializedKeyValuePairs.nonEmpty) {
-        ledgerStateOperations.writeState(serializedKeyValuePairs)
-      } else {
-        Future.unit
-      }
-      envelopedLogEntry = Envelope.enclose(entry)
-      _ = exporterWriteSet.foreach {
-        _ += entryId.toByteString -> envelopedLogEntry
-      }
-      index <- ledgerStateOperations.appendToLog(entryId.toByteString, envelopedLogEntry)
+      (serializedKeyValuePairs, envelopedLogEntry) <- inParallel(
+        Future(stateSerializationStrategy.serializeState(outputState)),
+        Future(Envelope.enclose(entry)),
+      )
+      (_, _, index) <- inParallel(
```

Comment:
Wouldn't `Future.traverse` yield the same?

Reply:
Yes, sort of, in a dynamic language. `Future.sequence` (or `Future.traverse(…)(identity)`) would have the same behavior, but we'd have to pass in a `Seq[Future[A]]` in and get a `Future[Seq[A]]` out. Because these values have different types, we'd end up with `Future[Seq[Any]]` (or something equally unsafe, such as `Future[Seq[Any with Product with Serializable]]`.

We need something that works on tuples, not sequences.

As the comment says, Cats and Scalaz both support doing something like this, but I'm too much of an `implicit` newbie to figure out how. (I tried. It was ugly.)

Reply:
Makes sense -- thanks for the additional color on your design choice.